File: ios/Sources/KlinikFinanceFeature/FinanceModel.swift
import Foundation
import KlinikAPI
import KlinikCore

public enum FinancePhase: Sendable, Equatable {
    case loading
    case loaded
    case notPermitted
    case failed(String)
}

public struct FinanceState: Sendable, Equatable {
    public var phase: FinancePhase = .loading
    public var currency: Currency = .turkishLira
    /// Nil means "everything"; the picker's first option.
    public var status: PaymentStatus?

    public var records: [FinanceRecord] = []
    public var nextCursor: String?
    public var outstanding: OutstandingReport?
    public var collections: CollectionReport?

    public var busyId: String?
    public var error: String?

    public init() {}

    public var hasMore: Bool { nextCursor != nil }
}

/**
 * What has been billed and what has been paid (spec M11).
 *
 * Three reads: the ledger, what is still owed with its ageing, and what came in
 * this month. Together they answer the two questions a clinic asks of a finance
 * screen — who owes us, and did the money arrive — and neither is answerable
 * from the record list alone.
 *
 * Nothing here computes money. Every figure is a string the server produced
 * from a decimal, and the client renders it. Arithmetic in the client would be
 * arithmetic in a second place, and the two would eventually disagree in front
 * of a patient.
 */
@MainActor
public final class FinanceModel {
    private let api: FinanceAPI
    private var state = FinanceState()

    public init(api: FinanceAPI) {
        self.api = api
    }

    public func currentState() -> FinanceState { state }

    public func load() async {
        async let page = optional {
            try await api.records(status: self.state.status, currency: self.state.currency)
        }
        async let outstanding = optional { try await api.outstanding(currency: self.state.currency) }
        async let collections = optional { try await self.thisMonthsCollections() }

        let loaded = await (page, outstanding, collections)

        state.records = loaded.0?.items ?? []
        state.nextCursor = loaded.0?.nextCursor
        state.outstanding = loaded.1
        state.collections = loaded.2

        state.phase = loaded.0 == nil && loaded.1 == nil && loaded.2 == nil
            ? .notPermitted
            : .loaded
    }

    public func loadMore() async {
        guard let cursor = state.nextCursor else { return }

        guard
            let page = try? await api.records(
                status: state.status,
                currency: state.currency,
                cursor: cursor
            )
        else { return }

        state.records += page.items
        state.nextCursor = page.nextCursor
    }

    public func choose(currency: Currency) async {
        state.currency = currency
        state.phase = .loading
        await load()
    }

    public func choose(status: PaymentStatus?) async {
        state.status = status
        state.phase = .loading
        await load()
    }

    /// Records a payment against one invoice and replaces that row with what
    /// the server returned — never with a locally adjusted balance.
    public func pay(
        recordId: String,
        amount: String,
        method: PaymentMethod,
        reference: String?
    ) async -> Bool {
        state.busyId = recordId
        state.error = nil

        defer { state.busyId = nil }

        do {
            let updated = try await api.recordPayment(
                recordId: recordId,
                amount: amount,
                method: method,
                reference: reference
            )

            if let index = state.records.firstIndex(where: { $0.id == recordId }) {
                state.records[index] = updated
            }

            // The ageing and the month's takings both moved.
            state.outstanding = try? await api.outstanding(currency: state.currency)
            state.collections = try? await thisMonthsCollections()

            return true
        } catch let error as APIError {
            state.error = L10n.message(for: error)
        } catch {
            state.error = L10n.string("error.server")
        }

        return false
    }

    /**
     * Reverses a payment (spec M11).
     *
     * The row stays and stops counting — the server's design, and the right
     * one: a payment that was entered and undone is part of what happened, and
     * a ledger that forgets its corrections is a ledger nobody can audit. The
     * reason is required for the same reason.
     */
    public func reverse(paymentId: String, reason: String) async -> Bool {
        let trimmed = reason.trimmingCharacters(in: .whitespacesAndNewlines)

        guard !trimmed.isEmpty else {
            state.error = L10n.string("finance.reverseNeedsReason")
            return false
        }

        state.busyId = paymentId
        state.error = nil

        defer { state.busyId = nil }

        do {
            let updated = try await api.reversePayment(paymentId, reason: trimmed)

            if let index = state.records.firstIndex(where: { $0.id == updated.id }) {
                state.records[index] = updated
            }

            state.outstanding = try? await api.outstanding(currency: state.currency)
            state.collections = try? await thisMonthsCollections()

            return true
        } catch let error as APIError {
            state.error = L10n.message(for: error)
        } catch {
            state.error = L10n.string("error.server")
        }

        return false
    }

    private func thisMonthsCollections() async throws -> CollectionReport {
        let calendar = Calendar.current
        let now = Date()
        let start = calendar.date(from: calendar.dateComponents([.year, .month], from: now)) ?? now

        return try await api.collections(from: start, to: now, currency: state.currency)
    }

    private func optional<T: Sendable>(_ work: @Sendable () async throws -> T) async -> T? {
        try? await work()
    }
}

public extension Amount {
    /// A figure with its symbol. Formatting only — the value is the server's.
    func display(_ currency: Currency) -> String {
        "\(currency.symbol)\(text)"
    }
}

File: android/feature/finance/src/main/kotlin/xyz/klinik/feature/finance/FinanceModel.kt
package xyz.klinik.feature.finance

import java.time.Instant
import java.time.ZoneId
import java.time.ZonedDateTime
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import xyz.klinik.network.ApiError
import xyz.klinik.network.CollectionReport
import xyz.klinik.network.Currency
import xyz.klinik.network.ExchangeRate
import xyz.klinik.network.FinanceApi
import xyz.klinik.network.FinanceRecord
import xyz.klinik.network.OutstandingReport
import xyz.klinik.network.PaymentMethod
import xyz.klinik.network.PaymentStatus
import xyz.klinik.network.UiText
import xyz.klinik.network.uiText

sealed interface FinancePhase {
    data object Loading : FinancePhase
    data object Loaded : FinancePhase

    /** Not a failure: an account without `finance.read` has no screen here. */
    data object NotPermitted : FinancePhase
}

data class FinanceState(
    val phase: FinancePhase = FinancePhase.Loading,
    val currency: Currency = Currency.TRY,
    /** Null means "everything" — the picker's first option. */
    val status: PaymentStatus? = null,
    val records: List<FinanceRecord> = emptyList(),
    val nextCursor: String? = null,
    val outstanding: OutstandingReport? = null,
    val collections: CollectionReport? = null,
    /**
     * The rates recorded for the month the collections cover.
     *
     * Empty is a real answer and the screen says so: it is why a total is
     * incomplete, and a reader looking at a figure that excludes two thousand
     * pounds deserves to know which.
     */
    val rates: List<ExchangeRate> = emptyList(),
    val busyId: String? = null,
    val error: UiText? = null,
) {
    val hasMore: Boolean get() = nextCursor != null
}

/**
 * What has been billed and what has been paid (spec M11).
 *
 * Four reads: the ledger, what is still owed with its ageing, what came in
 * this month, and the rates behind that month. Together they answer the two
 * questions a clinic asks of a finance screen — who owes us, and did the money
 * arrive — and neither is answerable from the record list alone.
 *
 * Nothing here computes money. Every figure is a string the server produced
 * from a decimal, and the client renders it. Arithmetic in the client would be
 * arithmetic in a second place, and the two would eventually disagree in front
 * of a patient.
 */
class FinanceModel(
    private val api: FinanceApi,
    private val clock: () -> Instant = Instant::now,
    private val zone: ZoneId = ZoneId.systemDefault(),
) {
    private val _state = MutableStateFlow(FinanceState())
    val state: StateFlow<FinanceState> = _state.asStateFlow()

    suspend fun load() = coroutineScope {
        val current = _state.value

        _state.value = current.copy(phase = FinancePhase.Loading)

        val (from, to) = thisMonth()

        val page = async { optional { api.records(current.status, current.currency) } }
        val outstanding = async { optional { api.outstanding(current.currency) } }
        val collections = async { optional { api.collections(from, to, current.currency) } }
        val rates = async { optional { api.rates(from, to) } }

        val ledger = page.await()
        val owed = outstanding.await()
        val taken = collections.await()

        _state.value = current.copy(
            phase = if (ledger == null && owed == null && taken == null) {
                FinancePhase.NotPermitted
            } else {
                FinancePhase.Loaded
            },
            records = ledger?.items.orEmpty(),
            nextCursor = ledger?.nextCursor,
            outstanding = owed,
            collections = taken,
            rates = rates.await().orEmpty(),
        )
    }

    suspend fun loadMore() {
        val cursor = _state.value.nextCursor ?: return
        val current = _state.value
        val page = optional { api.records(current.status, current.currency, cursor) } ?: return

        _state.value = _state.value.copy(
            records = _state.value.records + page.items,
            nextCursor = page.nextCursor,
        )
    }

    suspend fun choose(currency: Currency) {
        _state.value = _state.value.copy(currency = currency)
        load()
    }

    suspend fun choose(status: PaymentStatus?) {
        _state.value = _state.value.copy(status = status)
        load()
    }

    /**
     * Records money that arrived.
     *
     * The row is replaced with what the server returned rather than with a
     * locally adjusted balance: subtracting here would be the second place
     * this clinic does arithmetic, and the two would eventually disagree in
     * front of a patient.
     */
    suspend fun pay(
        recordId: String,
        amount: String,
        method: PaymentMethod,
        reference: String?,
    ): Boolean {
        _state.value = _state.value.copy(busyId = recordId, error = null)

        return try {
            val updated = api.recordPayment(recordId, amount, method, reference = reference)

            replace(updated)
            refreshTotals()

            true
        } catch (error: Throwable) {
            _state.value = _state.value.copy(busyId = null, error = messageFor(error))
            false
        }
    }

    /**
     * Reverses a payment (spec M11).
     *
     * The row stays and stops counting — the server's design, and the right
     * one: a payment that was entered and undone is part of what happened, and
     * a ledger that forgets its corrections is a ledger nobody can audit. The
     * reason is required for the same reason.
     */
    suspend fun reverse(paymentId: String, reason: String): Boolean {
        val trimmed = reason.trim()

        if (trimmed.isEmpty()) {
            _state.value = _state.value.copy(error = UiText.Key("finance.reverseNeedsReason"))
            return false
        }

        _state.value = _state.value.copy(busyId = paymentId, error = null)

        return try {
            val updated = api.reversePayment(paymentId, trimmed)

            replace(updated)
            refreshTotals()

            true
        } catch (error: Throwable) {
            _state.value = _state.value.copy(busyId = null, error = messageFor(error))
            false
        }
    }

    private fun replace(updated: FinanceRecord) {
        _state.value = _state.value.copy(
            records = _state.value.records.map { if (it.id == updated.id) updated else it },
            busyId = null,
        )
    }

    /** The ageing and the month's takings both moved. */
    private suspend fun refreshTotals() {
        val currency = _state.value.currency
        val (from, to) = thisMonth()

        _state.value = _state.value.copy(
            outstanding = optional { api.outstanding(currency) } ?: _state.value.outstanding,
            collections = optional { api.collections(from, to, currency) } ?: _state.value.collections,
        )
    }

    /** The window the collections and the rates both cover, so they read together. */
    private fun thisMonth(): Pair<String, String> {
        val moment = ZonedDateTime.ofInstant(clock(), zone)
        val start = moment.withDayOfMonth(1).toLocalDate().atStartOfDay(zone)

        return start.toInstant().toString() to moment.toInstant().toString()
    }

    private suspend fun <T> optional(work: suspend () -> T): T? =
        runCatching { work() }.getOrNull()

    private fun messageFor(error: Throwable): UiText =
        (error as? ApiError)?.uiText() ?: UiText.Key("error.server")
}
